src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -826,7 +876,66 @@ ss::future<abs_client::list_bucket_result> abs_client::do_list_objects(
 
 ss::future<result<abs_client::storage_account_info, error_outcome>>
 abs_client::get_account_info(ss::lowres_clock::duration timeout) {
-    return send_request(do_get_account_info(timeout), object_key{""});
+    if (_is_oauth) {
+        return send_request(
+          do_test_set_expiry_on_dummy_file(timeout), object_key{""});
+    } else {
+        return send_request(do_get_account_info(timeout), object_key{""});
+    }
+}
+
+ss::future<abs_client::storage_account_info>
+abs_client::do_test_set_expiry_on_dummy_file(
+  ss::lowres_clock::duration timeout) {
+    // since this is one-off operation at startup, it's easier to read directly
+    // cloud_storage_azure_container than to wire it in. this is ok because if
+    // we are in abs_client it means that the required property, like
+    // azure_container, are set
+    auto container_name
+      = config::shard_local_cfg().cloud_storage_azure_container.value();
+
+    if (unlikely(!container_name.has_value())) {
+        vlog(abs_log.error, "Failed to get azure container name from config");
+        throw std::runtime_error("cloud_storage_azure_container is not set");
+    }
+
+    auto bucket = bucket_name{container_name.value()};
+    auto test_file = object_key{set_expiry_test_file};
+
+    // try set expiry
+    auto set_expiry_header = _requestor.make_set_expiry_to_blob_request(
+      bucket, test_file, std::chrono::seconds{30});
+    if (!set_expiry_header) {
+        vlog(
+          abs_log.error,
+          "Failed to create set_expiry header: {}",
+          set_expiry_header.error());
+        throw std::system_error(set_expiry_header.error());
+    }
+
+    vlog(abs_log.trace, "send https request:\n{}", set_expiry_header.value());
+
+    auto response_stream = co_await _client.request(
+      std::move(set_expiry_header.value()), timeout);
+
+    co_await response_stream->prefetch_headers();
+    vassert(response_stream->is_header_done(), "Header is not received");
+    auto const& headers = response_stream->get_headers();
+
+    if (headers.result() == boost::beast::http::status::bad_request) {
```

Comment:
What's our strategy for testing that this code is correct? In production? :)

I'm worried that we might break this request by accident and infer the wrong value for whether HNS gets enabled.

Also, do they return a good error code in the body we can also check?

Reply:
> What's our strategy for testing that this code is correct? In production? :)

waiting for CDT to be updated to run on azure with managed identities, so for now local emulation + manual testing.

> I'm worried that we might break this request by accident and infer the wrong value for whether HNS gets enabled.

yes that's the reason for `cloud_storage_azure_hierarchical_namespace_enabled`

> Also, do they return a good error code in the body we can also check?

testing shows `x-ms-error-code: HierarchicalNamespaceNotEnabled` in the header and 
`<Error><Code>HierarchicalNamespaceNotEnabled</Code></Error>`
in the body, but it's not a code that's documented

we have to strategies: if we get a 400 but not the exact error code
1. log error (warn?) but assume no HNS
2. log error and stop redpanda from starting.

for 1, redpanda will start and if we are wrong there will be further errors

```
TRACE 2024-04-17 21:19:00,210 [shard 0:main] abs - abs_client.cc:840 - send https request:
GET /testingcontainer?restype=container&comp=list&prefix=cluster_metadata/1461eab4-668b-46bc-ae1e-97f188d46842/manifests/&delimiter=/ HTTP/1.1
Host: testingimds2ab.blob.core.windows.net
x-ms-version: 2023-01-03
Authorization: [secret]
 
ERROR 2024-04-17 21:19:00,240 [shard 0:main] abs - util.cc:108 - Unexpected error cloud_storage_clients::xml_parse_exception (Invalid state: parsing Last-Modified when not in Blob tag)
WARN  2024-04-17 21:19:00,240 [shard 0:main] cloud_storage - [fiber71~0|1|3599970ms] - remote.cc:1444 - ListObjectsV2 testingcontainer, unexpected error: cloud_storage_clients::error_outcome:1
WARN  2024-04-17 21:19:00,240 [shard 0:main] cluster - uploader.cc:377 - Manifest download failed in term 13: 2 0 List objects failedList objects failed
INFO  2024-04-17 21:19:00,240 [shard 0:main] cluster - uploader.cc:368 - Syncing cluster metadata manifest in term 13
``` 

in this case cluster/cloud_metadata/uploader.cc::upload_until_term_change() is not making progress because listing works differently under HNS

now, looking at the code it *seems* that the issue with listing might be salvageable with some tricks to tolerate the presence of keys for empty folders, and we *might* do the same for delete_object (the other method using adls endpoints to deal with directories), 

so the end result would be a log with errors and potentially orphaned objects, but redpanda should keep working, if we guessed wrongly that HNS is disabled.

I'll check ifthis makes sense and if we have test coverage for delete/listing